File: app/src/main/kotlin/com/youcoded/app/runtime/CommandProvider.kt
package com.youcoded.app.runtime

import com.youcoded.app.skills.LocalSkillProvider
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

// Mirrors desktop/src/main/command-provider.ts. Kept in sync manually —
// see docs/cc-dependencies.md entry "CC built-in command list" and
// docs/superpowers/specs/2026-04-21-command-drawer-search-commands-design.md.
//
// Last verified against Claude Code CLI v2.1.116 — 2026-04-21.
class CommandProvider(
  private val homeDir: File,
  private val skillProvider: LocalSkillProvider,
  private val getProjectCwd: () -> String?,
) {
  private var cache: JSONArray? = null

  fun invalidateCache() {
    cache = null
  }

  fun getCommands(): JSONArray {
    cache?.let { return it }

    val claudeDir = File(homeDir, ".claude")
    val entries = mutableListOf<JSONObject>()

    // Source 1: YouCoded-handled (hardcoded)
    entries.addAll(YOUCODED_COMMANDS)

    // Source 2: filesystem — user + project + plugin
    entries.addAll(scanCommandsFromDir(File(claudeDir, "commands"), null))
    getProjectCwd()?.let { cwd ->
      entries.addAll(scanCommandsFromDir(File(File(cwd), ".claude/commands"), null))
    }
    entries.addAll(scanPluginCommands(claudeDir))

    // Source 3: CC built-ins (hardcoded)
    entries.addAll(CC_BUILTIN_COMMANDS)

    // Dedup by name with precedence youcoded > filesystem > cc-builtin.
    // A command whose name matches an existing skill is dropped.
    val sourcePriority = mapOf("youcoded" to 0, "filesystem" to 1, "cc-builtin" to 2)
    val byName = linkedMapOf<String, JSONObject>()
    for (entry in entries) {
      val name = entry.getString("name")
      val existing = byName[name]
      if (existing == null ||
          sourcePriority[entry.getString("source")]!! < sourcePriority[existing.getString("source")]!!) {
        byName[name] = entry
      }
    }

    // Drop commands that collide with skills by name.
    val skillNames = mutableSetOf<String>()
    val skillList = skillProvider.getInstalled()
    for (i in 0 until skillList.length()) {
      val s = skillList.getJSONObject(i)
      skillNames.add("/" + s.getString("displayName"))
    }
    for (name in skillNames) byName.remove(name)

    val result = JSONArray()
    for (entry in byName.values) result.put(entry)
    cache = result
    return result
  }

  private fun scanCommandsFromDir(dir: File, pluginSlug: String?): List<JSONObject> {
    if (!dir.isDirectory) return emptyList()
    val out = mutableListOf<JSONObject>()
    for (file in dir.listFiles { f -> f.isFile && f.name.endsWith(".md") } ?: return emptyList()) {
      val stem = file.nameWithoutExtension
      val description = extractFrontmatterDescription(file.readText())
      val name = if (pluginSlug != null) "/$pluginSlug:$stem" else "/$stem"
      out.add(JSONObject().apply {
        put("name", name)
        put("description", description)
        put("source", "filesystem")
        put("clickable", true)
      })
    }
    return out
  }

  private fun scanPluginCommands(claudeDir: File): List<JSONObject> {
    val marketplaces = File(claudeDir, "plugins/marketplaces")
    if (!marketplaces.isDirectory) return emptyList()
    val out = mutableListOf<JSONObject>()
    for (mp in marketplaces.listFiles { f -> f.isDirectory } ?: return emptyList()) {
      val plugins = File(mp, "plugins")
      if (!plugins.isDirectory) continue
      for (plugin in plugins.listFiles { f -> f.isDirectory } ?: continue) {
        out.addAll(scanCommandsFromDir(File(plugin, "commands"), plugin.name))
      }
    }
    return out
  }

  private fun extractFrontmatterDescription(content: String): String {
    // Normalize CRLF → LF so Windows-authored files don't trip the fence
    // detection or leave \r in captured values (mirrors the desktop fix).
    val normalized = content.replace("\r\n", "\n")
    if (!normalized.startsWith("---")) return ""
    val end = normalized.indexOf("\n---", 3)
    if (end == -1) return ""
    val block = normalized.substring(3, end)
    val match = Regex("(?m)^\\s*description\\s*:\\s*(.+?)\\s*$").find(block) ?: return ""
    var value = match.groupValues[1].trim()
    if ((value.startsWith("\"") && value.endsWith("\"")) ||
        (value.startsWith("'") && value.endsWith("'"))) {
      value = value.substring(1, value.length - 1)
    }
    return value
  }

  companion object {
    private val YOUCODED_COMMANDS: List<JSONObject> = listOf(
      youcoded("/compact",  "Compact conversation with native spinner card"),
      youcoded("/clear",    "Clear conversation timeline with native marker"),
      youcoded("/reset",    "Clear conversation timeline with native marker"),
      youcoded("/new",      "Clear conversation timeline with native marker"),
      youcoded("/model",    "Open native model picker"),
      youcoded("/fast",     "Toggle fast mode"),
      youcoded("/effort",   "Open effort-level picker"),
      youcoded("/copy",     "Copy assistant response to clipboard"),
      youcoded("/resume",   "Open native Resume Browser"),
      youcoded("/config",   "Open Preferences popup"),
      youcoded("/settings", "Open Preferences popup"),
      youcoded("/cost",     "Show native Usage card"),
      youcoded("/usage",    "Show native Usage card"),
    )

    private val CC_BUILTIN_COMMANDS: List<JSONObject> = listOf(
      ccBuiltin("/help",            "Show Claude Code help"),
      ccBuiltin("/status",          "Show session, config, and auth status"),
      ccBuiltin("/permissions",     "Manage tool permissions"),
      ccBuiltin("/memory",          "Edit CLAUDE.md memory files"),
      ccBuiltin("/agents",          "Manage subagents"),
      ccBuiltin("/mcp",             "Manage MCP servers"),
      ccBuiltin("/plugin",          "Manage plugins"),
      ccBuiltin("/hooks",           "Manage hooks"),
      ccBuiltin("/doctor",          "Diagnose the installation"),
      ccBuiltin("/logout",          "Sign out of your Anthropic account"),
      ccBuiltin("/context",         "Show current context-window usage"),
      ccBuiltin("/review",          "Review a pull request"),
      ccBuiltin("/security-review", "Review pending changes for security issues"),
      ccBuiltin("/init",            "Initialize a CLAUDE.md file"),
      ccBuiltin("/extra-usage",     "Show detailed usage data"),
      ccBuiltin("/heapdump",        "Dump a heap snapshot"),
      ccBuiltin("/insights",        "Show session insights"),
      ccBuiltin("/team-onboarding", "Team setup flow"),
    )

    private fun youcoded(name: String, description: String) = JSONObject().apply {
      put("name", name); put("description", description)
      put("source", "youcoded"); put("clickable", true)
    }

    private fun ccBuiltin(name: String, description: String) = JSONObject().apply {
      put("name", name); put("description", description)
      put("source", "cc-builtin"); put("clickable", false)
      put("disabledReason", "Please run $name in Terminal View.")
    }
  }
}

File: desktop/src/main/command-provider.ts
import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs';
import type { CommandEntry, SkillEntry } from '../shared/types';
import { YOUCODED_COMMANDS, expandWithAliases } from './youcoded-commands';
import { CC_BUILTIN_COMMANDS } from './cc-builtin-commands';
import { scanCommandsFromDir, scanPluginCommandsDir } from './command-scanner';

// Precedence for name collisions: YouCoded (native UI) beats filesystem,
// which beats CC built-in. A command whose name matches an existing skill
// is dropped entirely (avoids double-listing skill-backed CC commands like
// /review, /init, /security-review that ship as skills).
export function mergeCommandSources(
  youcoded: CommandEntry[],
  filesystem: CommandEntry[],
  ccBuiltin: CommandEntry[],
  skills: SkillEntry[],
): CommandEntry[] {
  // Build the skill-name set (names normalized with leading slash for
  // comparison against CommandEntry.name).
  const skillNames = new Set<string>(skills.map((s) => `/${s.displayName}`));

  const byName = new Map<string, CommandEntry>();
  // Insert in reverse precedence so higher-precedence sources overwrite.
  for (const entry of ccBuiltin)  byName.set(entry.name, entry);
  for (const entry of filesystem) byName.set(entry.name, entry);
  for (const entry of youcoded)   byName.set(entry.name, entry);

  // Drop anything that collides with a skill name.
  for (const skillName of skillNames) byName.delete(skillName);

  return Array.from(byName.values());
}

// Stateful provider. Caches the merged list for the session lifetime,
// invalidated via `invalidateCache()` when plugin install/uninstall changes
// the filesystem. Mirrors the LocalSkillProvider caching pattern.
//
// getCommands() is async because LocalSkillProvider.getInstalled() is async
// (even though its body is synchronous — the declaration is `async`, so it
// always returns a Promise). Awaiting the skills list is load-bearing: the
// skill-name dedup step in mergeCommandSources requires a real SkillEntry[]
// array, not a Promise.
export class CommandProvider {
  private cache: CommandEntry[] | null = null;
  private getSkills: () => Promise<SkillEntry[]>;
  private getProjectCwd: () => string | null;

  constructor(
    getSkills: () => Promise<SkillEntry[]>,
    getProjectCwd: () => string | null,
  ) {
    this.getSkills = getSkills;
    this.getProjectCwd = getProjectCwd;
  }

  invalidateCache(): void {
    this.cache = null;
  }

  async getCommands(): Promise<CommandEntry[]> {
    if (this.cache) return this.cache;

    const home = os.homedir();
    const claudeDir = path.join(home, '.claude');

    const youcoded = expandWithAliases(YOUCODED_COMMANDS);

    // Filesystem: user + project + plugin commands
    const user = scanCommandsFromDir(path.join(claudeDir, 'commands'));
    const cwd = this.getProjectCwd();
    const project = cwd ? scanCommandsFromDir(path.join(cwd, '.claude', 'commands')) : [];
    const plugin = scanAllPluginCommandDirs(claudeDir);
    const filesystem = [...user, ...project, ...plugin];

    const skills = await this.getSkills();
    this.cache = mergeCommandSources(youcoded, filesystem, CC_BUILTIN_COMMANDS, skills);
    return this.cache;
  }
}

// Walk `~/.claude/plugins/marketplaces/*/plugins/*/commands/` and collect
// every plugin's namespaced commands.
function scanAllPluginCommandDirs(claudeDir: string): CommandEntry[] {
  const marketplacesRoot = path.join(claudeDir, 'plugins', 'marketplaces');
  const out: CommandEntry[] = [];

  let marketplaces: fs.Dirent[];
  try {
    marketplaces = fs.readdirSync(marketplacesRoot, { withFileTypes: true });
  } catch {
    return out;
  }

  for (const mp of marketplaces) {
    if (!mp.isDirectory()) continue;
    const pluginsRoot = path.join(marketplacesRoot, mp.name, 'plugins');
    let plugins: fs.Dirent[];
    try {
      plugins = fs.readdirSync(pluginsRoot, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const plugin of plugins) {
      if (!plugin.isDirectory()) continue;
      const pluginDir = path.join(pluginsRoot, plugin.name);
      out.push(...scanPluginCommandsDir(pluginDir, plugin.name));
    }
  }
  return out;
}
